pinot-core/src/main/java/org/apache/pinot/core/indexsegment/mutable/MutableSegmentImpl.java

```diff
@@ -472,10 +476,8 @@ public SegmentMetadata getSegmentMetadata() {
   public ColumnDataSource getDataSource(String columnName) {
     FieldSpec fieldSpec = _schema.getFieldSpecFor(columnName);
     if (fieldSpec.isVirtualColumn()) {
-      VirtualColumnContext virtualColumnContext =
-          new VirtualColumnContext(NetUtil.getHostnameOrAddress(), getSegmentName(), columnName, _numDocsIndexed + 1);
-      VirtualColumnProvider virtualColumnProvider =
-          VirtualColumnProviderFactory.buildProvider(_schema.getFieldSpecFor(columnName).getVirtualColumnProvider());
+      VirtualColumnContext virtualColumnContext = new VirtualColumnContext(fieldSpec, _numDocsIndexed);
```

Comment:
I'm not sure why `_numDocsIndexed + 1` is used - it's causing `MutableSegmentImplTest` to fail after adding virtual columns to mutable segments.

Reply:
Hmm.. Does it work before your change? Can you find out why?

Reply:
Previously virtual columns were not added to the mutable segments. I added them in this diff.

In `MutableSegmentImplTest` there is a test validating the total number of documents for each column (physical and virtual) in the mutable segments and immutable segment that both read from the same avro file. It was only checking physical columns before. After I add the virtual columns to the mutable segment, the test failed due to one extra row in the mutable segment, likely caused by this ` + 1`

Reply:
I think this `+1` is a bug. (numDocs is misused as index)

Reply:
Ok, all the usage seems to refer to document count, which indeed should be numDocsIndexed and not numDocsIndexed+1. I think if we had more tests that actually look up inverted index, etc, on doc Id, it would have revealed the earlier bug. 
I did a quick test of running LLCClusterIntegrationTest and ran some queries against it 
SELECT count(*) from mytable where $segmentName = "provided the last consuming segment here"
SELECT max($docId) from mytable where $segmentName = "provided the last consuming segment here"

I did this on current master as well as this fork. It looks like the responses come out correctly.

As an aside, we don't get virtual columns  on any realtime segments other than consuming ones. It feels strange. I think we stopped adding virtual column to completed data sometime ago...